internal/xds/xdsdependencymanager/xds_dependency_manager.go

```diff
@@ -0,0 +1,241 @@
+
+	ldsResourceName       string
+	listenerWatcher       *listenerWatcher
+	currentListenerUpdate *xdsresource.ListenerUpdate
+
+	rdsResourceName    string
+	currentRouteConfig xdsresource.RouteConfigUpdate
+	routeConfigWatcher *routeConfigWatcher
+	currentVirtualHost *xdsresource.VirtualHost
+
+	logger *grpclog.PrefixLogger
+}
+
+// New creates a new dependency manager that manages all the xds resources
+// including LDS, RDS, CDS and EDS and sends update once we have all the
+// resources and sends an error when we get error in listener or route
+// resources.
+func New(listenername, dataplaneAuthority string, xdsClient xdsclient.XDSClient, watcher ConfigWatcher) *DependencyManager {
+	// Builds the dependency manager and starts the listener watch.
+	dm := &DependencyManager{
+		ldsResourceName:    listenername,
+		dataplaneAuthority: dataplaneAuthority,
+		xdsClient:          xdsClient,
+		watcher:            watcher,
+	}
+	dm.logger = prefixLogger(dm)
+
+	// Initialize the serializer used to synchronize the updates from the xDS
+	// client.
+	ctx, cancel := context.WithCancel(context.Background())
+	dm.serializer = grpcsync.NewCallbackSerializer(ctx)
+	dm.serializerCancel = cancel
+
+	// Start the listener watch. Listener watch will start the other resource
+	// watches as needed.
+	dm.listenerWatcher = newListenerWatcher(listenername, dm)
+	return dm
+}
+
+// Close closes the dependency manager and stops all the watchers registered.
+func (m *DependencyManager) Close() {
+	// Cancel the context passed to the serializer and wait for any scheduled
+	// callbacks to complete. Canceling the context ensures that no new
+	// callbacks will be scheduled.
+	m.serializerCancel()
+	<-m.serializer.Done()
+
+	if m.listenerWatcher != nil {
+		m.listenerWatcher.stop()
+	}
+	if m.routeConfigWatcher != nil {
+		m.routeConfigWatcher.stop()
+	}
+}
+
+// ConfigWatcher is notified of the XDSConfig resource updates and errors that
+// are received by the xDS client from the management server. It only receives a
+// XDSConfig update after all the xds resources have been received.
+type ConfigWatcher interface {
+	// OnUpdate is invoked by the dependency manager to provide a new,
+	// validated xDS configuration to the watcher.
+	OnUpdate(xdsresource.XDSConfig)
+
+	// OnError is invoked when an error is received in listener or route
+	// resource. This includes cases where:
+	//  - The listener or route resource watcher reports a resource error.
+	//  - The received listener resource is a socket listener, not an API listener - TODO :  This is not yet implemented, tracked here #8114
+	//  - The received route configuration does not contain a virtual host
+	//  matching the channel's default authority.
+	OnError(error)
+}
+
+func (m *DependencyManager) maybeSendUpdate() {
+	if m.logger.V(2) {
+		m.logger.Infof("Sending update to watcher: Listener: %v, RouteConfig: %v", pretty.ToJSON(m.currentListenerUpdate), pretty.ToJSON(m.currentRouteConfig))
```

Comment:
We've had many performance problems with using `pretty.JSON` for printing structs. I would recommend using `%+v` or some other native formatting directive instead.

Another thing to consider is also whether the xDS resolver also outputs this log. If so, we don't want the same information being repeated twice.